.sandbox/provision/lifecycle-asg/common.sh

```diff
@@ -0,0 +1,100 @@
+#!/bin/bash
+
+function redirect_stdout() {
+    exec 3>&1
+    exec 1>&2
+}
+
+function restore_stdout() {
+    exec 1>&3
+    exec 3>&-
+}
+
+function get_volume_id() {
+    volume_info="$(aws ec2 describe-volumes --filters Name=tag:SandboxID,Values=$SANDBOX_ID)"
+    jq -r '.Volumes[0].VolumeId' <<< $volume_info
+}
+
+function get_instance_id() {
+    instances_info=$(aws ec2 describe-instances --filters Name=tag:SandboxID,Values=$SANDBOX_ID)
+    jq -r '.Reservations[0].Instances[0].InstanceId' <<< $instances_info
+}
+
+# claim_instance_from_asg claims an instance from ASG if there is no existing one claimed.
+function claim_instance_from_asg() {
+    # chceck is there any running instance already
+    instance_with_sandbox_id="$(aws ec2 describe-instances --filters Name=tag:SandboxID,Values=$SANDBOX_ID Name=instance-state-name,Values=running)"
+    if [[ $(jq '.Reservations[0].Instances | length' <<< "$instance_with_sandbox_id") -eq 0 ]]; then 
+        INSTANCE_IDS="$(aws autoscaling describe-auto-scaling-groups --auto-scaling-group-name $ASG_NAME --no-paginate --query "AutoScalingGroups[].Instances[].InstanceId" --output text)"
+        INSTANCE_ID="$(echo $INSTANCE_IDS | awk '{print $1}')"
+        _="$(aws autoscaling detach-instances --instance-ids $INSTANCE_ID --auto-scaling-group-name $ASG_NAME --no-should-decrement-desired-capacity)"
+        aws ec2 create-tags --resources $INSTANCE_ID --tags Key=Sandbox,Value=$SANDBOX_NAME --tags Key=SandboxID,Value=$SANDBOX_ID
```

Comment:
Edge case: L30 succeeded, and the script is killed for any reason before running L31. How to handle the instance leak?

Reply:
Regarding the tags, I don't think `Sandbox` (for sandbox name) is necessary, and it may introduce confusion because sandbox name is not unique (expanding multiple lifecycles, e.g. create a sandbox named `foo`, and delete it and then create another one named `foo`).

Reply:
I think what we learnt from Employ case is that `Sandbox` is needed and it would be very helpful during debugging.

Reply:
Regarding the leaked instance, it has the following properties:
- It does not have SandboxID tag
- It has ASG suggested tag `sandbox-asg:true`
- It is not linked to any ASG.

We can document this somewhere to indicate the ASG must include the below tag:

`sandbox-asg:true`